### ISSUE TITLE
[FEATURE REQ] Buffered Upload for Append Blobs

### ISSUE THREAD

**Issue by abhikt48** [Azure/azure-sdk-for-java#20393]
We are migrating blob component from old SDK a`zure-storage-5.5.0` to new SDK `azure-storage-blob-12.10.0`.  
**Old SDK** - We used `CloudAppendBlob.upload(inputStream, -1)` to append blob with **-1** length size. We are getting input as stream, so it will be difficult to get exact size of input. 
As per Java docs - `InputStream.available()` can provide estimate size only which is not reliable.
So we are using **-1** for stream object with old SDK which is working fine.

**New SDK** - I tried samething with new SDK to append blob and upload blob but I am getting below `UnexpectedLengthException `error, please find below details. As per java docs, we need to provide exact length, which is not possible for Stream object, and we can't provide exact length of a Stream object.

I observed same issue in other blob client as well - `BlobClient.upload(stream, -1, true);`

Can I request you to please provide option to upload blob with unknow size. Please find below details and let me know for any issues.

**New SDK Code for Append blob**
`AppendBlobClient.appendBlock(blobAsInputStream, -1);`

**Code ScreenShot**
![image](https://user-images.githubusercontent.com/22897837/113724239-0ed0a300-96ea-11eb-845e-bb82bcac65c5.png)

**Error StackTrace**
```
com.azure.core.exception.UnexpectedLengthException: Request body emitted 71 bytes, more than the expected -1 bytes.
	at com.azure.storage.common.Utility.lambda$convertStreamToByteBuffer$3(Utility.java:292) ~[azure-storage-common-12.10.0.jar:?]
	at reactor.core.publisher.FluxPeek$PeekSubscriber.onComplete(FluxPeek.java:243) ~[reactor-core-3.3.12.RELEASE.jar:3.3.12.RELEASE]
	at reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.drain(FluxConcatMap.java:359) ~[reactor-core-3.3.12.RELEASE.jar:3.3.12.RELEASE]
	at reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.onComplete(FluxConcatMap.java:268) ~[reactor-core-3.3.12.RELEASE.jar:3.3.12.RELEASE]
	at reactor.core.publisher.FluxMap$MapSubscriber.onComplete(FluxMap.java:136) ~[reactor-core-3.3.12.RELEASE.jar:3.3.12.RELEASE]
	at reactor.core.publisher.Operators.complete(Operators.java:135) ~[reactor-core-3.3.12.RELEASE.jar:3.3.12.RELEASE]
	at reactor.core.publisher.FluxEmpty.subscribe(FluxEmpty.java:41) ~[reactor-core-3.3.12.RELEASE.jar:3.3.12.RELEASE]
	at reactor.core.publisher.InternalFluxOperator.subscribe(InternalFluxOperator.java:62) ~[reactor-core-3.3.12.RELEASE.jar:3.3.12.RELEASE]
	at reactor.core.publisher.FluxDefer.subscribe(FluxDefer.java:54) ~[reactor-core-3.3.12.RELEASE.jar:3.3.12.RELEASE]
	at reactor.core.publisher.FluxSubscribeOn$SubscribeOnSubscriber.run(FluxSubscribeOn.java:187) ~[reactor-core-3.3.12.RELEASE.jar:3.3.12.RELEASE]
	at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:84) ~[reactor-core-3.3.12.RELEASE.jar:3.3.12.RELEASE]
	at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:37) ~[reactor-core-3.3.12.RELEASE.jar:3.3.12.RELEASE]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_242]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) ~[?:1.8.0_242]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) ~[?:1.8.0_242]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_242]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_242]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_242]
	Suppressed: java.lang.Exception: #block terminated with an error
		at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:99) ~[reactor-core-3.3.12.RELEASE.jar:3.3.12.RELEASE]
		at reactor.core.publisher.Mono.block(Mono.java:1685) ~[reactor-core-3.3.12.RELEASE.jar:3.3.12.RELEASE]
		at com.azure.storage.common.implementation.StorageImplUtils.blockWithOptionalTimeout(StorageImplUtils.java:131) ~[azure-storage-common-12.10.0.jar:?]
		at com.azure.storage.blob.specialized.AppendBlobClient.appendBlockWithResponse(AppendBlobClient.java:227) ~[azure-storage-blob-12.10.0.jar:?]
		at com.azure.storage.blob.specialized.AppendBlobClient.appendBlock(AppendBlobClient.java:192) ~[azure-storage-blob-12.10.0.jar:?]
```

**Java SDK**
![image](https://user-images.githubusercontent.com/22897837/113722045-f2cc0200-96e7-11eb-9233-b4d5e7690c54.png)

**Old SDK**
![image](https://user-images.githubusercontent.com/22897837/113721856-c9ab7180-96e7-11eb-9578-8cffc3d4e981.png)

**Setup (please complete the following information if applicable):**
OS: window
IDE : eclipse
Version of the Library used: `azure-storage-blob-12.10.0`

**Comment by gapra-msft** [comment]
Hi @abhikt48 
Thank you for posting this feature request. 

As you can see from the docs, the AppendBlob upload does require that you know the size of the input before hand. 

We will discuss this request with the rest of the team and determine if we can take this on as a feature request and appropriately schedule it in our backlog

**Comment by dkm199** [comment]
+1 for this feature.

**Comment by jaschrep-msft** [comment]
Two different issues are being brought up here.

`BlobClient.upload(stream, -1, true)` is being brought upon the side, and the fix for that is far simpler. This has to do with our sync to async conversions and how InputStreams are currently converted. This upgrade is currently being looked into and is tracked under https://github.com/Azure/azure-sdk-for-java/issues/20567.

`AppendBlobClient.appendBlock(stream, -1)` is an entirely separate issue. While `BlobClient` is utilizing a buffering strategy to allow for simpler APIs, parallelization, etc., `AppendBlobClient`, `BlockBlobClient`, and `PageBlobClient` do not currently buffer your data. Without buffering data, we cannot determine the data length, which is needed when putting requests on the wire. Accepting a length of -1 here would require significant changes.

Further, the legacy SDK method being mentioned is not even a single append method, but a multi-block append method that (when given -1) would buffer chunks from the `InputStream` and append them each until the end of the InputStream was reached (this is equivalent to how `BlobClient` provides helpful functionality on top of `BlockBlobClient` in this SDK). Without this, the maximum size of InputStream would be the maximum size of an append block (currently 4MB).

Editing this issue title to be a feature request for a buffered upload on `AppendBlobClient` (and `AppendBlobAsyncClient`)

@abhikt48 to be clear, the feature we would end up writing to support this scenario would be, in one way or another, a convenience where we loop read 4MB chunks from the `InputStream` into a buffer and then call `appendBlock()` for you using the buffered data length as the length argument. You could write a simple version of this to unblock yourself if the functionality is critical.

**Comment by glammers1** [comment]
@abhikt48 hi! did you manage to implement your own version of _buffered uploads_ for the Append client? I'm trying it with a pretty straightforward approach creating chunks of 4MB as follows:

```java
public Flux<AppendBlobItem> appendBlock(String containerName, String blobName, InputStream data) {
	return FluxUtil.toFluxByteBuffer(data, MAX_BYTES_APPEND_BLOCK).flatMap((chunk) -> {
		log.info("Appending chunk of length (in bytes) [{}] to container [{}] and blob [{}]", chunk.limit(), containerName, blobName);
		return this.getAppendBlobAsyncClient(containerName, blobName).appendBlock(Flux.just(chunk), chunk.limit());
	});
}
```

Where `MAX_BYTES_APPEND_BLOCK` is the 4MB that `appendBlock` method allows from the SDK. 

**The file is uploaded but when I try to download it the data is corrupted**. I'm downloading it from the Azure portal UI directly so I don't have a `download` method written by myself that could be the guilty. I would say there is something going on inside the `appendBlock` method when sending chunks... **This is only happening with the Async client**. The same approach is working fine with the Sync client (the sync client receives an`InputStream` instead of `Flux<ByteBuffer>` as input). 

@jaschrep-msft pinging you as well in case you have any tips to implement this. By the way, if I read from a file into `Flux<ByteBuffer>`, create chunks of 4MB and write to a local file, the file is not corrupted. That's why I think the way I'm creating the chunks is correct and there might be something corrupting the data when using `appendBlock`.

This is the local example:

```java
DataInputStream reader = new DataInputStream(new FileInputStream("myfile.pptx"));
var fluxByteBuffer = FluxUtil.toFluxByteBuffer(reader, 4000000); // 4MB
FluxUtil.writeFile(fluxByteBuffer, AsynchronousFileChannel.open(Path.of("myfile_v2.pptx"), StandardOpenOption.CREATE_NEW, StandardOpenOption.WRITE)).subscribe();
```

**Comment by jaschrep-msft** [comment]
@glammers1 Sorry for the delay in response. `FluxUtil.writeFile()` has reactive code to guarantee no concurrent writes. You code does not have that guarantee in place. You could be getting into a race condition and can confirm this by observing the corrupted data to look for blocks of the observed size being out of order.

If that is not the nature of your corruption, please open a separate bug report rather than continuing on this feature request.

**Comment by glammers1** [comment]
> @abhikt48 hi! did you manage to implement your own version of _buffered uploads_ for the Append client? I'm trying it with a pretty straightforward approach creating chunks of 4MB as follows:
> 
> ```java
> public Flux<AppendBlobItem> appendBlock(String containerName, String blobName, InputStream data) {
> 	return FluxUtil.toFluxByteBuffer(data, MAX_BYTES_APPEND_BLOCK).flatMap((chunk) -> {
> 		log.info("Appending chunk of length (in bytes) [{}] to container [{}] and blob [{}]", chunk.limit(), containerName, blobName);
> 		return this.getAppendBlobAsyncClient(containerName, blobName).appendBlock(Flux.just(chunk), chunk.limit());
> 	});
> }
> ```
> 
> Where `MAX_BYTES_APPEND_BLOCK` is the 4MB that `appendBlock` method allows from the SDK.
> 
> **The file is uploaded but when I try to download it the data is corrupted**. I'm downloading it from the Azure portal UI directly so I don't have a `download` method written by myself that could be the guilty. I would say there is something going on inside the `appendBlock` method when sending chunks... **This is only happening with the Async client**. The same approach is working fine with the Sync client (the sync client receives an`InputStream` instead of `Flux<ByteBuffer>` as input).
> 
> @jaschrep-msft pinging you as well in case you have any tips to implement this. By the way, if I read from a file into `Flux<ByteBuffer>`, create chunks of 4MB and write to a local file, the file is not corrupted. That's why I think the way I'm creating the chunks is correct and there might be something corrupting the data when using `appendBlock`.
> 
> This is the local example:
> 
> ```java
> DataInputStream reader = new DataInputStream(new FileInputStream("myfile.pptx"));
> var fluxByteBuffer = FluxUtil.toFluxByteBuffer(reader, 4000000); // 4MB
> FluxUtil.writeFile(fluxByteBuffer, AsynchronousFileChannel.open(Path.of("myfile_v2.pptx"), StandardOpenOption.CREATE_NEW, StandardOpenOption.WRITE)).subscribe();
> ```
> 
> I'm using files of 10MB for my tests.

I found the issue with my code above some time ago but wanting to write down here what was happening. It has nothing to do with the `appendBlock` call but with the usage of `flatMap`.

The `flatMap` operation in Flux does not guarantee any specific order of processing for the elements. I changed it to `concatMap` which preserves the order. With that change, the data is not getting corrupted anymore.
